s3tests_boto3/functional/test_s3select.py

```diff
@@ -0,0 +1,410 @@
+import boto3
+import botocore.session
+from botocore.exceptions import ClientError
+from botocore.exceptions import ParamValidationError
+from nose.tools import eq_ as eq
+from nose.plugins.attrib import attr
+from nose.plugins.skip import SkipTest
+import isodate
+import email.utils
+import datetime
+import threading
+import re
+import pytz
+from collections import OrderedDict
+import requests
+import json
+import base64
+import hmac
+import hashlib
+import xml.etree.ElementTree as ET
+import time
+import operator
+import nose
+import os
+import string
+import random
+import socket
+import ssl
+from email.header import decode_header
+
+from .utils import assert_raises
+from .utils import generate_random
+from .utils import _get_status_and_error_code
+from .utils import _get_status
+
+from .policy import Policy, Statement, make_json_policy
+
+from . import (
+    get_client,
+    get_prefix,
+    get_unauthenticated_client,
+    get_bad_auth_client,
+    get_v2_client,
+    get_new_bucket,
+    get_new_bucket_name,
+    get_new_bucket_resource,
+    get_config_is_secure,
+    get_config_host,
+    get_config_port,
+    get_config_endpoint,
+    get_main_aws_access_key,
+    get_main_aws_secret_key,
+    get_main_display_name,
+    get_main_user_id,
+    get_main_email,
+    get_main_api_name,
+    get_alt_aws_access_key,
+    get_alt_aws_secret_key,
+    get_alt_display_name,
+    get_alt_user_id,
+    get_alt_email,
+    get_alt_client,
+    get_tenant_client,
+    get_tenant_iam_client,
+    get_tenant_user_id,
+    get_buckets_list,
+    get_objects_list,
+    get_main_kms_keyid,
+    get_secondary_kms_keyid,
+    nuke_prefixed_buckets,
+    )
+
+import boto
+import boto.s3.connection
+import sys
+import random
+from botocore.client import Config
+
+region_name = ''
+
+def get_connection():
+    conn = boto.connect_s3(
+        aws_access_key_id = get_main_aws_access_key(),
+        aws_secret_access_key = get_main_aws_secret_key(),
+        host = get_config_host(),
+        port = get_config_port(),
+        is_secure=False,               # uncomment if you are not using ssl
+        calling_format = boto.s3.connection.OrdinaryCallingFormat(),
+        )
+
+    return conn
+
+def create_csv_object_for_datetime(rows,columns):
+        result = ""
+        for i in range(rows):
+            row = "";
+            for y in range(columns):
+                row = row + "{}{:02d}{:02d}-{:02d}{:02d}{:02d},".format(random.randint(0,100)+1900,random.randint(1,12),random.randint(1,28),random.randint(0,23),random.randint(0,59),random.randint(0,59),);
+            result += row + "\n"
+
+        return result
+
+def create_random_csv_object(rows,columns,col_delim=",",record_delim="\n",csv_schema=""):
```

Comment:
what about string columns?

Reply:
this function create outputs, for testing the integration between {rows X columns} and s3-select query processing.
there are tests for checking CSV different use-cases.

Reply:
the random generator, create only numeric columns, can we have tests against csv with string columns as well. this is unrelated to csv use-cases, this would verify string functions, operators, where clause on string columns etc.

Reply:
by default all tokens are string (for CSV), in order to get it as number it must be cast first.
indeed the current framework generate only numbers, but there are actually strings.
there are tests including where clause with substr functions.

```python
    res_s3select_substr = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select min(int(_2)),max(int(_2)) from stdin where substr(_2,1,1) == "1"')).replace("\n","")
```

Reply:
> by default all tokens are string (for CSV), in order to get it as number it must be cast first.
> indeed the current framework generate only numbers, but there are actually strings.
> there are tests including where clause with substr functions.
> 
> ```python
>     res_s3select_substr = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select min(int(_2)),max(int(_2)) from stdin where substr(_2,1,1) == "1"')).replace("\n","")
> ```

what about string operators (e.g. str+str)?

Reply:
+ is not implemented for string.